From 7505ab0acdfd2341a8935535d926d074b5d20dde Mon Sep 17 00:00:00 2001
From: David Sterba <dsterba@suse.cz>
Date: Fri, 15 Nov 2013 14:21:02 +0100
Subject: [PATCH 02/16] feature: skip-quoted

Tunable for number of lines of quoted text that are displayed before the
unquoted text after 'skip to quoted' command (S).

Signed-off-by: David Sterba <dsterba@suse.cz>
---
 PATCHES                |   1 +
 README.skip-quoted     |  47 +++++++++++++++++++
 doc/manual.xml.head    | 121 +++++++++++++++++++++++++++++++++++++++++++++++++
 doc/muttrc.skip-quoted |  10 ++++
 globals.h              |   1 +
 init.h                 |   6 +++
 pager.c                |   8 ++--
 7 files changed, 190 insertions(+), 4 deletions(-)
 create mode 100644 README.skip-quoted
 create mode 100644 doc/muttrc.skip-quoted

diff --git a/PATCHES b/PATCHES
index 58ef6c9..b8b631e 100644
--- a/PATCHES
+++ b/PATCHES
@@ -1,3 +1,4 @@
+patch-skip-quoted-neo-git
 patch-progress-neo-git
 patch-sidebar-neo-git
 patch-compress-neo-git
diff --git a/README.skip-quoted b/README.skip-quoted
new file mode 100644
index 0000000..fe611d1
--- /dev/null
+++ b/README.skip-quoted
@@ -0,0 +1,47 @@
+Skip-Quoted Patch
+=================
+
+    Leave some context visible
+
+Patch
+-----
+
+    To check if Mutt supports "skip-quoted", look for "patch-skip-quoted" in
+    the mutt version.
+
+    Dependencies
+    * mutt-1.6.2
+
+Introduction
+------------
+
+    When viewing an email, the '<skip-to-quoted>' function (by default the 'S'
+    key) will scroll past any quoted text. Sometimes, a little context is
+    useful.
+
+    By setting the '$skip_quoted_offset' variable, you can select how much of
+    the quoted text is left visible.
+
+Variables
+---------
+
+    | Name                 | Type   | Default |
+    |----------------------|--------|---------|
+    | 'skip_quoted_offset' | number | 0       |
+
+See Also
+--------
+
+    * NeoMutt project
+
+Known Bugs
+----------
+
+    None
+
+Credits
+-------
+
+    * David Sterba <dsterba@suse.cz>
+    * Richard Russon <rich@flatcap.org>
+
diff --git a/doc/manual.xml.head b/doc/manual.xml.head
index a58759c..3901190 100644
--- a/doc/manual.xml.head
+++ b/doc/manual.xml.head
@@ -10648,6 +10648,127 @@ <sect1 id="sidebar">
   </sect2>
 </sect1>
 
+<sect1 id="skip-quoted-patch">
+  <title>Skip-Quoted Patch</title>
+  <subtitle>Leave some context visible</subtitle>
+
+  <sect2 id="skip-quoted-patch2">
+    <title>Patch</title>
+
+    <para>
+      To check if Mutt supports <quote>skip-quoted</quote>, look for
+      <quote>patch-skip-quoted</quote> in the mutt version.
+      See: <xref linkend="mutt-patches"/>.
+    </para>
+
+    <itemizedlist>
+      <title>Dependencies:</title>
+      <listitem><para>mutt-1.6.2</para></listitem>
+    </itemizedlist>
+
+    <para>This patch is part of the <ulink url="http://www.neomutt.org/">NeoMutt Project</ulink>.</para>
+  </sect2>
+
+  <sect2 id="skip-quoted-intro">
+    <title>Introduction</title>
+
+    <para>
+      When viewing an email, the
+      <literal>&lt;skip-to-quoted&gt;</literal> function (by default the
+      <literal>S</literal> key) will scroll past any quoted text.
+      Sometimes, a little context is useful.
+    </para>
+
+    <para>
+      By setting the <literal>$skip_quoted_offset</literal> variable, you
+      can select how much of the quoted text is left visible.
+    </para>
+  </sect2>
+
+  <sect2 id="skip-quoted-variables">
+    <title>Variables</title>
+    <table id="table-skip-quoted-variables">
+      <title>Skip-Quoted Variables</title>
+      <tgroup cols="3">
+        <thead>
+          <row>
+            <entry>Name</entry>
+            <entry>Type</entry>
+            <entry>Default</entry>
+          </row>
+        </thead>
+        <tbody>
+          <row>
+            <entry><literal>skip_quoted_offset</literal></entry>
+            <entry>number</entry>
+            <entry>0</entry>
+          </row>
+        </tbody>
+      </tgroup>
+    </table>
+  </sect2>
+
+<!--
+  <sect2 id="skip-quoted-functions">
+    <title>Functions</title>
+    <para>None</para>
+  </sect2>
+
+  <sect2 id="skip-quoted-commands">
+    <title>Commands</title>
+    <para>None</para>
+  </sect2>
+
+  <sect2 id="skip-quoted-colors">
+    <title>Colors</title>
+    <para>None</para>
+  </sect2>
+
+  <sect2 id="skip-quoted-sort">
+    <title>Sort</title>
+    <para>None</para>
+  </sect2>
+-->
+
+  <sect2 id="skip-quoted-muttrc">
+    <title>Muttrc</title>
+
+<screen>
+<emphasis role="comment"># Example Mutt config file for the 'skip-quoted' patch.
+ 
+# The 'S' (skip-quoted) command scrolls the pager past the quoted text (usually
+# indented with '&gt; '.  Setting 'skip_quoted_offset' leaves some lines of quoted
+# text on screen for context.
+ 
+# Show three quoted lines before the reply</emphasis>
+set skip_quoted_offset = 3
+ 
+<emphasis role="comment"># vim: syntax=muttrc</emphasis>
+</screen>
+  </sect2>
+
+  <sect2 id="skip-quoted-see-also">
+    <title>See Also</title>
+
+    <itemizedlist>
+      <listitem><para><ulink url="http://www.neomutt.org/">NeoMutt Project</ulink></para></listitem>
+    </itemizedlist>
+  </sect2>
+
+  <sect2 id="skip-quoted-known-bugs">
+    <title>Known Bugs</title>
+    <para>None</para>
+  </sect2>
+
+  <sect2 id="skip-quoted-credits">
+    <title>Credits</title>
+    <itemizedlist>
+    <listitem><para>David Sterba <email>dsterba@suse.cz</email></para></listitem>
+    <listitem><para>Richard Russon <email>rich@flatcap.org</email></para></listitem>
+    </itemizedlist>
+  </sect2>
+</sect1>
+
 <sect1 id="trash-folder">
   <title>Trash Folder Patch</title>
   <subtitle>Automatically move <quote>deleted</quote> emails to a trash bin</subtitle>
diff --git a/doc/muttrc.skip-quoted b/doc/muttrc.skip-quoted
new file mode 100644
index 0000000..cd13ed4
--- /dev/null
+++ b/doc/muttrc.skip-quoted
@@ -0,0 +1,10 @@
+# Example Mutt config file for the 'skip-quoted' patch.
+ 
+# The 'S' (skip-quoted) command scrolls the pager past the quoted text (usually
+# indented with '> '.  Setting 'skip_quoted_offset' leaves some lines of quoted
+# text on screen for context.
+ 
+# Show three quoted lines before the reply
+set skip_quoted_offset = 3
+ 
+# vim: syntax=muttrc
diff --git a/globals.h b/globals.h
index d81166b..66a1da1 100644
--- a/globals.h
+++ b/globals.h
@@ -211,6 +211,7 @@ WHERE short ReflowWrap;
 WHERE short SaveHist;
 WHERE short SendmailWait;
 WHERE short SleepTime INITVAL (1);
+WHERE short SkipQuotedOffset;
 WHERE short TimeInc;
 WHERE short Timeout;
 WHERE short Wrap;
diff --git a/init.h b/init.h
index 0b565a8..3e53f72 100644
--- a/init.h
+++ b/init.h
@@ -2862,6 +2862,12 @@ struct option_t MuttVars[] = {
   ** replacing ``%s'' with the supplied string.
   ** For the default value, ``joe'' would be expanded to: ``~f joe | ~s joe''.
   */
+  { "skip_quoted_offset", DT_NUM, R_NONE, UL &SkipQuotedOffset, 0 },
+  /*
+  ** .pp
+  ** Lines of quoted text that are displayed before the unquoted text after
+  ** "skip to quoted" command (S)
+  */
   { "sleep_time",	DT_NUM, R_NONE, UL &SleepTime, 1 },
   /*
   ** .pp
diff --git a/pager.c b/pager.c
index e8b2b04..226fe15 100644
--- a/pager.c
+++ b/pager.c
@@ -2300,11 +2300,11 @@ search_next:
 	  int dretval = 0;
 	  int new_topline = topline;
 
-	  while ((new_topline < lastLine ||
+	  while (((new_topline + SkipQuotedOffset) < lastLine ||
 		  (0 == (dretval = display_line (fp, &last_pos, &lineInfo,
 			 new_topline, &lastLine, &maxLine, M_TYPES | (flags & M_PAGER_NOWRAP),
 			 &QuoteList, &q_level, &force_redraw, &SearchRE))))
-		 && lineInfo[new_topline].type != MT_COLOR_QUOTED)
+		 && lineInfo[new_topline + SkipQuotedOffset].type != MT_COLOR_QUOTED)
 	    new_topline++;
 
 	  if (dretval < 0)
@@ -2313,11 +2313,11 @@ search_next:
 	    break;
 	  }
 
-	  while ((new_topline < lastLine ||
+	  while (((new_topline + SkipQuotedOffset) < lastLine ||
 		  (0 == (dretval = display_line (fp, &last_pos, &lineInfo,
 			 new_topline, &lastLine, &maxLine, M_TYPES | (flags & M_PAGER_NOWRAP),
 			 &QuoteList, &q_level, &force_redraw, &SearchRE))))
-		 && lineInfo[new_topline].type == MT_COLOR_QUOTED)
+		 && lineInfo[new_topline + SkipQuotedOffset].type == MT_COLOR_QUOTED)
 	    new_topline++;
 
 	  if (dretval < 0)
-- 
2.7.4

